docs/design/node_taint_eviction.md

```diff
@@ -42,51 +48,42 @@ automaticReplacementOptions:
-    for node, _ := range nodeMap {
-        // Check if node has a taint for the minimum duration in that case we will set the value to true.
-    }
-
-    podMap := internal.CreatePodMap(cluster, pods)
-    for _, processGroup := range status.ProcessGroups {
-        pod, podExists := podMap[processGroup.ProcessGroupID]
-        if !podExists {
-            continue
-        }
-
-        tainted, nodeFound := nodeTaintedMap[pod.Spec.NodeName]
-        if !nodeFound {
-            continue
-        }
-
-        processGroup.UpdateCondition(fdbv1beta2.RunningOnTaintedNode, tainted, status.ProcessGroups, processGroup.ProcessGroupID)
-    }
-}
+In the above configuration, it takes at least 3000 seconds (1200 + 1800) for the operator to replace a Pod on a Node with taint key `example.com/disconnected`, because it takes `durationInSeconds`, i.e., 1200 seconds, to mark the Pod to be replaced and it takes another `taintReplacementTimeSeconds`, i.e., 1800 seconds, to start replacing the Pod.
+
+**Requirement 3.** The feature lists the taint status for process groups. This allows users to have a holistic view of which Pods in a FDB cluster are running on tainted Nodes and being replaced.
+
+### Implementation
+We need to make the following four type of changes:
+
+**CRD change.** We need to extend the `automaticReplacementOptions` with the new fields `taints` and `taintReplacementTimeSeconds`
+
+**Detect Node taint and change Pod condition.**  We can piggy-back the existing `validateProcessGroup()` function in `update_status` reconciler to detect if a Pod is running on a tainted Node and mark the Pod's taint-related condition accordingly.
+
+A Pod's Node can be listed by
 ```
+node = &corev1.Node{}
+err := r.Get(ctx, client.ObjectKey{Name: pod.Spec.NodeName}, node)
+```
+
+**Status of Pod taint-related condition.** We will add two  [ProcessGroupConditionType](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/api/v1beta2/foundationdbcluster_types.go#L673): `NodeTaintDetected`, which represents a Pod's Node is tainted but not long enough for operator to replace it, and `NodeTaintReplacing`, which represents a Pod whose Node has been tainted for longer-than-configured time and the operator should replace it.
+
+**Replace Pods with `NodeTaintReplacing` conditioin.** We can piggy-back existing replacement logic to replace Pods with `NodeTaintReplacing` condition. We can do this by adding `NodeTaintReplacing` to [`conditionsThatNeedReplacement` list](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/d38d6bc7abf7764215976b75d581c933280389f8/api/v1beta2/foundationdbcluster_types.go#L65-L67)
+
+
+## Limitations
+The operator will only detect tainted Nodes when its reconciliation is triggered periodically. This may increase the delay of detecting and replacing Pods on tainted Nodes.
```

Comment:
```suggestion
## Limitations

The operator will only detect tainted Nodes when its reconciliation is triggered. This may increase the delay of detecting and replacing Pods on tainted Nodes.
```

The periodically statement is wrong here. The operator will also (potentially) detected this condition if a reconciliation was triggered by an unrelated event, saying that the tainted nodes are only detected by the periodic reconciliation is wrong. We can add a statement here that node changes are not triggering any reconciliation.